iiif_prezi3/helpers/add_thumbnail.py

```diff
@@ -21,5 +22,66 @@ def add_thumbnail(self, image_url, **kwargs):
         self.thumbnail.append(new_thumbnail)
         return new_thumbnail
 
+    def create_thumbnail_from_iiif(self, url, preferred_width=500, **kwargs):
+        """Adds an image thumbnail to a manifest or canvas based on a IIIF service.
+
+        Args:
+            url (str): An HTTP URL which points at a IIIF Image response.
+            preferred_width (int, optional): the preferred width of the thumbnail.
+            **kwargs (): see ResourceItem.
+
+        Returns:
+            new_thumbnail (ResourceItem): The updated list of thumbnails, including the newly-created one.
+        """
+        image_response = ResourceItem(id=url, type='Image')
+        image_info = image_response.set_hwd_from_iiif(url)
+        context = image_info.get('@context', '')
+
+        if 'sizes' in image_info:
+            best_fit_size = min(
+                image_info['sizes'], key=lambda size: abs(size["width"] - preferred_width)
```

Comment:
Would be good to talk through this (so i can understand how it works :-))

Reply:
Ideally get the closest image which is bigger than requested.

Reply:
I've refactored this to be:

```
best_fit_size = max(
                (size for size in image_info['sizes'] if size["width"] <= preferred_width),
                key=lambda size: size["width"]
            )
```

The max function takes a list of dicts that represents the sizes prop of an image response and finds the best dict where the width of the dict is less than or equal to the preferred width that is specified. Origninally, this used `abs()` to find the closest dict, but I've changed this to `max()` to make sure the width is never larger than the `width` that is preferred.  Since we're dealing with a list of dicts and not a list of insts, we need to be able to compare in the max function just the width of the dict but ultimately return the best dict based on our criteria. This is where the `key` comes in.  This tells the max function to only look at the width key when comparing each dictionary to preferred width.

There is arguably a more readable and understandable way to do this, but I'm struggling to come up with a better way.